feat: execute actions v2 on events

stebenz opened the pull request:
# Which Problems Are Solved

ZITADEL is not able to run configured event executions whenever an event is written to the event store. 

# How the Problems Are Solved

Addition of handling logic of created events based on already existing components. There is a separate process which handles the calling of the defined Targets, and this process only need the necessary information like the content of the event and the necessary list of Targets to call.

# Additional Changes

Additional testing in integration, to garantie processing of events.

# Additional Context

Closes #7246

adlerhurst commented on internal/execution/handlers.go:
```diff
@@ -0,0 +1,177 @@
+			aggList[aggType] = []eventstore.EventType{eventstore.EventType(eventType)}
+		} else {
+			found := false
+			for _, aggEventType := range aggEventTypes {
+				if aggEventType == eventstore.EventType(eventType) {
+					found = true
+				}
+			}
```
i would use slices.Contains

stebenz replied:
True, will change it

adlerhurst commented on internal/execution/handlers.go:
```diff
@@ -0,0 +1,177 @@
+	parts := strings.Split(s, ".")
+	groups := make([]string, len(parts))
+	groupBase := ""
+	for i, part := range parts {
+		if groupBase == "" {
+			groupBase = part
+		} else {
+			groupBase = groupBase + "." + part
+		}
+
+		if groupBase == s {
+			groups[i] = groupBase
+		} else {
+			groups[i] = groupBase + ".*"
+		}
+	}
```
This code feels very complex for the task it is solving, isn't there a function in the strings package which makes the code more readable?

stebenz replied:
Not that I know of, the thing is here with the best match, for the query there are as many possible values as there are elements in the group definition. 
For example for `user.human.mfa.init.skipped`:
- all
- user
- user.human
- user.human.mfa
- user.human.mfa.init
- user.human.mfa.init.skipped

stebenz replied:
There is no direct function for it, but I reworked it with strings.join

adlerhurst commented on internal/execution/target.go:
```diff
@@ -0,0 +1,85 @@
+	statusCode int,
+	respBody interface{},
+) (string, func(), func() bool) {
+	server := &server{
+		called: false,
+	}
```
server overwrites the server struct type, if no defaults use `new(server)`

stebenz replied:
Removed the file, as the TestServerCall is only used in tests

adlerhurst commented on internal/execution/worker.go:
```diff
@@ -0,0 +1,84 @@
+
+	targets, err := TargetsFromRequest(job.Args)
+	if err != nil {
+		return river.JobCancel(errors.New("targets not valid"))
```
add the original error

stebenz replied:
Added

adlerhurst commented on internal/execution/worker.go:
```diff
@@ -0,0 +1,84 @@
+
+	_, err = CallTargets(ctx, targets, exec_repo.ContextInfoFromRequest(job.Args))
+	if err != nil {
+		return river.JobCancel(err)
```
JobCancel stops retries so I wouldn't cancel the job if a target wasn't available once.

stebenz replied:
The `CallTargets` function includes all error handling for the list of targets, which means you only get an error returned if the `InterruptOnError` is set. That's why I think to cancel the Job is an indicator of the failed execution, or should that be handled differently?

stebenz replied:
That's specifically how it is designed, there are no retries, and an error is only returned if the target is defined with `interruptOnError`. 
And with what do you disagree? I mean, should it be handled differently in your opinion? Do you disagree with any remark I made there? Would you handle the job differently?

stebenz replied:
Yes I do disagree there, failure count does not necessarily mean that you need to retry, just that acknowledge that the call has failed. 
That you for example could handle if you create multiple users and every call failed, that you deactivate the execution.
Or are the failed queue jobs not visible anymore?

adlerhurst replied:
ok please document this behavior so that it's clear for users of actions and update the acceptance criteria that we do not retry and therefore behave different than projections.

adlerhurst commented on internal/execution/queries.go:
```diff
@@ -0,0 +1,44 @@
+	}
+}
+
+func (q *ExecutionsQueries) ActiveInstancesWithFeatureFlag(ctx context.Context) []string {
```
either my IDE and global search are stupid or the code is not used

stebenz replied:
You're right, seems as I had quite a confusion of the queries, cleanup the old interface

adlerhurst reviewed (changes requested):
great that you removed the dead code i missed, nice catch.

stebenz commented on cmd/setup/setup.go:
```diff
@@ -529,4 +529,9 @@ func initProjections(
 		logging.WithFields("name", p.String()).OnError(err).Fatal("migration failed")
 	}
+
+	for _, p := range execution.Projections() {
```
True, removed as it was left from the projection before.

stebenz commented on cmd/mirror/projections.go:
```diff
@@ -220,7 +222,14 @@ func projections(
+		config.Executions,
+		queries,
+		es.EventTypes(),
 		nil,
```
True, removed.

stebenz commented on cmd/mirror/projections.go:
```diff
@@ -295,6 +304,13 @@ func execProjections(ctx context.Context, instances <-chan string, failedInstanc
 			continue
 		}
+
+		err = execution.ProjectInstance(ctx)
```
True, removed.

stebenz commented on internal/execution/worker.go:
```diff
@@ -0,0 +1,84 @@
+}
+
+func (w *Worker) Timeout(*river.Job[*exec_repo.Request]) time.Duration {
+	return w.config.TransactionDuration
```
The timeout of the targets is managed in the `CallTargets` function, this is configurable that the transaction should also have a maximum timeout, as theoretically you could have an infinite amount of targets.

adlerhurst replied:
can you please add a comment that we ignore the sum to have a circuit breaker

stebenz replied:
There is never a mention that the sum of the timeouts is important? 
I don't get why we should mention this anywhere if in the API is a clear comment with `if the target doesn't respond before this timeout expires, the the connection is closed and the action fails`

stebenz replied:
As this function is defined by the river.Worker interface and there is a comment:
```
// Timeout is the maximum amount of time the job is allowed to run before
	// its context is cancelled. A timeout of zero (the default) means the job
	// will inherit the Client-level timeout. A timeout of -1 means the job's
	// context will never time out.
```

Still confused about the reasoning here, as it seems to also to be inconsistent to the notification worker, which uses the same logic regarding the error handling.
Added the comment to resolve this topic.

stebenz commented on cmd/start/start.go:
```diff
@@ -665,3 +681,14 @@ func checkExisting(values []string) func(string) bool {
 	}
 }
+
+func checkExistingGroup(values []string) func(string) bool {
```
Changed

adlerhurst commented on cmd/defaults.yaml:
```diff
@@ -450,13 +450,25 @@ Notifications:
+  # multi binary / pod setup and allowing only certain executables to process the events.
+  Workers: 1 # ZITADEL_EXECUTIONS_WORKERS
+  # The maximum duration a job can do it's work before it is considered as failed.
+  # This maximum duration is prioritised in case that the sum of the target's timeouts is higher,
```
```suggestion
  # This maximum duration is prioritized in case that the sum of the target's timeouts is higher,
```

stebenz commented on internal/execution/handlers.go:
```diff
@@ -0,0 +1,163 @@
+	for _, eventType := range u.eventTypes {
+		aggType := u.aggregateTypeFromEventType(eventstore.EventType(eventType))
+		aggEventTypes, ok := aggList[aggType]
+		if !ok {
+			aggList[aggType] = []eventstore.EventType{eventstore.EventType(eventType)}
+		} else if !slices.Contains(aggEventTypes, eventstore.EventType(eventType)) {
+			aggList[aggType] = append(aggList[aggType], eventstore.EventType(eventType))
+		}
```
Hm true, just added the conditional to clearly define the item of the map if not existing.

stebenz commented on internal/execution/handlers.go:
```diff
@@ -0,0 +1,163 @@
+		}
+	}
+
+	aggReducers := make([]handler.AggregateReducer, len(aggList))
+	i := 0
+	for aggType, aggEventTypes := range aggList {
+		eventReducers := make([]handler.EventReducer, len(aggEventTypes))
+		for j, eventType := range aggEventTypes {
+			eventReducers[j] = handler.EventReducer{
+				Event:  eventType,
+				Reduce: u.reduce,
+			}
+		}
+		aggReducers[i] = handler.AggregateReducer{
+			Aggregate:     aggType,
+			EventReducers: eventReducers,
+		}
+		i++
+	}
```
Are you sure about this? This would mean that with every aggregate the list is resized, even if we now the length.

stebenz replied:
True, changed

stebenz commented on internal/execution/handlers.go:
```diff
@@ -0,0 +1,163 @@
+func groupsFromEventType(s string) []string {
+	parts := strings.Split(s, ".")
+	groups := make([]string, len(parts))
+	for i := range parts {
+		groups[i] = strings.Join(parts[:i+1], ".")
+		if i < len(parts)-1 {
+			groups[i] += ".*"
+		}
+	}
+	// sort to end up with the most specific group first
+	slices.SortFunc(groups, func(a, b string) int {
+		return strings.Compare(a, b) * -1
+	})
```
True, changed it, but not sure if it makes it easier to understand

adlerhurst commented:
@stebenz the branch is out of date please solve that before i do a review

stebenz commented:
> @stebenz the branch is out of date please solve that before i do a review

v3.x is merged back into this branch

adlerhurst commented on internal/execution/handlers.go:
```diff
@@ -0,0 +1,158 @@
+}
+
+func groupsFromEventType(s string) []string {
+	parts := strings.Split(s, ".")
+	partsLen := len(parts)
+	groups := make([]string, partsLen)
+	for i := range parts {
+		groups[(partsLen-1)-i] = strings.Join(parts[:i+1], ".")
+		if i < len(parts)-1 {
+			groups[(partsLen-1)-i] += ".*"
+		}
+	}
+	return groups
```
sorry for commenting again I just remembered that there is the reverse function
```suggestion
    parts := strings.Split(s, ".")
	groups := make([]string, len(parts))
	for i := range parts {
		groups[i] = strings.Join(parts[:i+1], ".")
		if i < len(parts)-1 {
			groups[i] += ".*"
		}
	}
    slices.Reverse(groups)
    return groups
```

stebenz replied:
Overlooked, my bad, but good find, didn't now of the reverse function.
Changed.

adlerhurst reviewed (commented):
there is one left comment i didn't found an answer. and please update the issue (check acceptance criteria, change where needed and add the documentation of the event actions to the issue for the action docs if not done)

stebenz commented:
> there is one left comment i didn't found an answer. and please update the issue (check acceptance criteria, change where needed and add the documentation of the event actions to the issue for the action docs if not done)

Changed the issues, and added the missed comment as change.
Examples are added to https://github.com/zitadel/zitadel/issues/9456
